packages/core-app-api/src/apis/implementations/auth/pinniped/Pinniped.test.ts

```diff
@@ -0,0 +1,227 @@
+ * limitations under the License.
+ */
+
+import Pinniped from './Pinniped';
+import { ConfigReader } from '@backstage/config';
+import MockOAuthApi from '../../OAuthRequestApi/MockOAuthApi';
+import { UrlPatternDiscovery } from '../../DiscoveryApi';
+import { OAuth2Session } from '../oauth2';
+
+const createSession = jest.fn();
+const refreshSession = jest.fn();
+
+jest.mock('../../../../lib/AuthConnector/PinnipedAuthConnector', () => ({
+  ...(jest.requireActual(
+    '../../../../lib/AuthConnector/PinnipedAuthConnector',
+  ) as any),
+  PinnipedAuthConnector: class {
+    createSession = createSession;
+    refreshSession = refreshSession;
+  },
+}));
+
+describe('Pinniped', () => {
+  beforeEach(() => {
+    createSession.mockReset();
+    refreshSession.mockReset();
+  });
+
+  it('provisions fresh token', async () => {
+    refreshSession.mockRejectedValue(
+      new Error('Auth refresh request failed, Unauthorized'),
+    );
+
+    const createdSession: OAuth2Session = {
+      providerInfo: {
+        idToken: 'It is a cluster scope IdToken',
+        accessToken: '',
+        scopes: new Set(),
+        expiresAt: undefined,
+      },
+      profile: {
+        email: undefined,
+        displayName: undefined,
+        picture: undefined,
+      },
+    };
+
+    createSession.mockResolvedValue(createdSession);
+
+    const pinnipedAuth = Pinniped.create({
+      configApi: new ConfigReader(undefined),
+      oauthRequestApi: new MockOAuthApi(),
+      discoveryApi: UrlPatternDiscovery.compile(
+        'http://backstage.test/api/{{ pluginId }}',
+      ),
+    });
+
+    const idToken = await pinnipedAuth.getClusterScopedIdToken('myAudience');
+
+    expect(idToken).toBe('It is a cluster scope IdToken');
+  });
+
+  it('The session is renewed when there is less than 5 minutes until expiration', async () => {
+    refreshSession.mockRejectedValue(
+      new Error('Auth refresh request failed, Unauthorized'),
+    );
+
+    const expiresAt = new Date();
+    expiresAt.setSeconds(expiresAt.getSeconds() + 60 * 5 - 1);
+
+    const createdSession: OAuth2Session = {
+      providerInfo: {
+        idToken: 'It is a cluster scope IdToken',
+        accessToken: '',
+        scopes: new Set(),
+        expiresAt,
+      },
+      profile: {
+        email: undefined,
+        displayName: undefined,
+        picture: undefined,
+      },
+    };
+
+    createSession.mockResolvedValue(createdSession);
+
+    const pinnipedAuth = Pinniped.create({
+      configApi: new ConfigReader(undefined),
+      oauthRequestApi: new MockOAuthApi(),
+      discoveryApi: UrlPatternDiscovery.compile(
+        'http://backstage.test/api/{{ pluginId }}',
+      ),
+    });
+
+    const idToken = await pinnipedAuth.getClusterScopedIdToken('myAudience');
+
+    expect(idToken).toBe('It is a cluster scope IdToken');
+    expect(createSession).toHaveBeenCalledTimes(1);
+    expect(refreshSession).toHaveBeenCalledTimes(1);
```

Comment:
I tend to prefer unit tests that make assertions only once; I think these intervening assertions are basically covered in the happy path `provisions fresh token` spec above?
```suggestion
    await pinnipedAuth.getClusterScopedIdToken('myAudience');
```